File: code/analysis/migration/state_income_tax_savings.R
gc()
rm(list = ls())

library(duckdb)
library(dplyr)
library(tidyr)

# Connect to database
con <- dbConnect(duckdb::duckdb(), dbdir = "../../../data/database.db", read_only = TRUE)

county_returns_agi <- dbGetQuery(con, "SELECT * FROM returns_county_agi WHERE year = 2017")
state_returns      <- dbGetQuery(con, "SELECT * FROM returns_state WHERE year = 2017")

dbDisconnect(con, shutdown = TRUE)

# ---- Stub midpoints (in $000s, matching SOI amount units) ----
# County data has stubs 1, 3, 4, 5, 6, 8 (2→3 and 7→8 merged on import)
stub_midpoints <- c("1" = 0, "3" = 25, "4" = 62.5, "5" = 87.5, "6" = 150, "8" = 300)

# ---- Step 1: County-level stats ----
# A18425 (state_and_local_income_taxes_amount) is only observed for itemizers,
# so avg_state_inc_tax_pp understates total state taxes paid. However,
# cross-state differences in the effective rate remain informative.

county_stats <- county_returns_agi |>
    mutate(stub_midpoint = stub_midpoints[as.character(agi_stub)]) |>
    reframe(
        origin_state          = first(statefips),
        county_name           = first(countyname),
        avg_agi_pp            = sum(stub_midpoint * returns, na.rm = TRUE) /
                                sum(returns, na.rm = TRUE),
        avg_state_inc_tax_pp  = sum(state_and_local_income_taxes_amount, na.rm = TRUE) /
                                sum(returns, na.rm = TRUE),
        .by = county
    ) |>
    mutate(origin_effective_rate = avg_state_inc_tax_pp / avg_agi_pp) |>
    rename(county_fips = county)

cat("Counties:", nrow(county_stats), "\n")

# ---- Step 2: State-level effective income tax rates ----
# State data has stubs 1–8 (2 and 7 not pre-merged); merge to match county stubs.
state_rates <- state_returns |>
    mutate(agi_stub = case_when(
        agi_stub == 2L ~ 3L,
        agi_stub >= 8L ~ 8L,
        TRUE ~ agi_stub
    )) |>
    reframe(
        state_and_local_income_taxes_amount = sum(state_and_local_income_taxes_amount, na.rm = TRUE),
        returns = sum(returns, na.rm = TRUE),
        .by = c(state, agi_stub)
    ) |>
    mutate(stub_midpoint = stub_midpoints[as.character(agi_stub)]) |>
    reframe(
        dest_effective_rate = sum(state_and_local_income_taxes_amount, na.rm = TRUE) /
                              sum(stub_midpoint * returns, na.rm = TRUE),
        dest_inc_tax_pp     = sum(state_and_local_income_taxes_amount, na.rm = TRUE) /
                              sum(returns, na.rm = TRUE),
        .by = state
    ) |>
    rename(dest_state = state)

cat("States:", nrow(state_rates), "\n")

# ---- Step 3: Cross-join counties × states, exclude same-state moves ----
result <- county_stats |>
    cross_join(state_rates) |>
    filter(dest_state != origin_state) |>
    mutate(
        rate_diff             = origin_effective_rate - dest_effective_rate,
        tax_savings_thousands = avg_agi_pp * rate_diff
    ) |>
    select(
        county_fips,
        county_name,
        origin_state,
        dest_state,
        avg_agi_pp_thousands         = avg_agi_pp,
        avg_state_inc_tax_pp_thousands = avg_state_inc_tax_pp,
        origin_effective_rate,
        dest_effective_rate,
        rate_diff,
        tax_savings_thousands
    ) |>
    arrange(county_fips, dest_state)

cat("Output rows:", nrow(result), "\n")

# ---- Verification checks ----
cat("\n--- Verification ---\n")

# State-level effective rates (sorted)
cat("\nEffective state income tax rates (destination states), sorted:\n")
state_rates |>
    arrange(desc(dest_effective_rate)) |>
    print(n = 15)

# Sanity: CA county savings moving to TX
ca_to_tx <- result |>
    filter(origin_state == 6, dest_state == 48) |>
    arrange(desc(tax_savings_thousands))
cat("\nTop CA counties: savings moving to TX ($000s):\n")
print(head(ca_to_tx |> select(county_fips, county_name, avg_agi_pp_thousands,
                               origin_effective_rate, rate_diff, tax_savings_thousands), 5))

# Sanity: TX county savings moving to CA (should be negative)
tx_to_ca <- result |>
    filter(origin_state == 48, dest_state == 6)
cat("\nTX counties: savings moving to CA ($000s) — expect negative:\n")
print(head(tx_to_ca |> select(county_fips, county_name, avg_agi_pp_thousands,
                               origin_effective_rate, rate_diff, tax_savings_thousands), 5))

# ---- Write output ----
output_path <- "../../../data/county_to_state_tax_savings_2017.csv"
write.csv(result, output_path, row.names = FALSE)
cat("\nSaved to:", output_path, "\n")

File: code/analysis/migration/state_income_tax_savings.py
"""
County-to-state income tax savings, 2017.

For the average taxpayer in each county, computes:
  (a) tax_savings_thousands : annual state income taxes saved by moving to a
      given state (positive = saves money)
  (b) rate_diff             : difference in effective state income tax rate
      (positive = destination is cheaper)

Data source: IRS Statistics of Income (SOI) county and state returns, 2017.
  - state_and_local_income_taxes_amount (A18425): state income taxes deducted
    on federal returns. Observed only for itemizers, so this understates total
    state taxes paid. Cross-state *differences* in effective rates remain
    informative since the bias applies consistently across states.
  - Income estimated via AGI stub midpoints (no total AGI in database).

Output: /app/data/county_to_state_tax_savings_2017.csv
"""

import duckdb
import pandas as pd

DB_PATH = "../../../data/database.db"
OUT_PATH = "../../../data/county_to_state_tax_savings_2017.csv"

# AGI stub midpoints in $000s (matches SOI amount units)
# County data has stubs 1, 3, 4, 5, 6, 8 (2→3 and 7→8 merged on import)
STUB_MIDPOINTS = {1: 0.0, 3: 25.0, 4: 62.5, 5: 87.5, 6: 150.0, 8: 300.0}

# ---- Load data ----
con = duckdb.connect(DB_PATH, read_only=True)
county = con.execute(
    "SELECT county, statefips, countyname, agi_stub, returns, "
    "       state_and_local_income_taxes_amount "
    "FROM returns_county_agi WHERE year = 2017"
).df()
state = con.execute(
    "SELECT state, agi_stub, returns, state_and_local_income_taxes_amount "
    "FROM returns_state WHERE year = 2017"
).df()
con.close()

print(f"County rows: {len(county):,}")
print(f"State rows:  {len(state):,}")

# ---- County-level stats ----
county["stub_midpoint"] = county["agi_stub"].map(STUB_MIDPOINTS)

county_stats = (
    county.groupby(["county", "statefips", "countyname"])
    .apply(
        lambda g: pd.Series({
            "avg_agi_pp":           (g["stub_midpoint"] * g["returns"]).sum() / g["returns"].sum(),
            "avg_state_inc_tax_pp": g["state_and_local_income_taxes_amount"].sum() / g["returns"].sum(),
        }),
        include_groups=False,
    )
    .reset_index()
)
county_stats["origin_effective_rate"] = (
    county_stats["avg_state_inc_tax_pp"] / county_stats["avg_agi_pp"]
)
county_stats = county_stats.rename(columns={
    "county":    "county_fips",
    "statefips": "origin_state",
    "countyname": "county_name",
})
print(f"Counties: {len(county_stats):,}")

# ---- State-level effective income tax rates ----
# Merge stubs 2→3 and 7+→8 to match county stub coding
state["agi_stub"] = state["agi_stub"].apply(
    lambda s: 3 if s == 2 else (8 if s >= 8 else s)
)
state_agg = (
    state.groupby(["state", "agi_stub"])[
        ["state_and_local_income_taxes_amount", "returns"]
    ]
    .sum()
    .reset_index()
)
state_agg["stub_midpoint"] = state_agg["agi_stub"].map(STUB_MIDPOINTS)

state_rates = (
    state_agg.groupby("state")
    .apply(
        lambda g: pd.Series({
            "dest_effective_rate": (
                g["state_and_local_income_taxes_amount"].sum()
                / (g["stub_midpoint"] * g["returns"]).sum()
            ),
            "dest_inc_tax_pp": (
                g["state_and_local_income_taxes_amount"].sum() / g["returns"].sum()
            ),
        }),
        include_groups=False,
    )
    .reset_index()
    .rename(columns={"state": "dest_state"})
)
print(f"States: {len(state_rates):,}")

# ---- Cross-join counties × states, exclude same-state moves ----
county_stats["_key"] = 1
state_rates["_key"] = 1
result = county_stats.merge(state_rates, on="_key").drop(columns="_key")
result = result[result["dest_state"] != result["origin_state"]].copy()

result["rate_diff"] = result["origin_effective_rate"] - result["dest_effective_rate"]
result["tax_savings_thousands"] = result["avg_agi_pp"] * result["rate_diff"]

result = result.rename(columns={
    "avg_agi_pp":           "avg_agi_pp_thousands",
    "avg_state_inc_tax_pp": "avg_state_inc_tax_pp_thousands",
})
result = result[
    [
        "county_fips", "county_name", "origin_state", "dest_state",
        "avg_agi_pp_thousands", "avg_state_inc_tax_pp_thousands",
        "origin_effective_rate", "dest_effective_rate",
        "rate_diff", "tax_savings_thousands",
    ]
].sort_values(["county_fips", "dest_state"])

print(f"Output rows: {len(result):,}")

# ---- Verification ----
print("\n--- Effective rates by destination state (sorted descending) ---")
print(
    state_rates.sort_values("dest_effective_rate", ascending=False)
    .to_string(index=False)
)

print("\n--- Top CA counties: savings moving to TX (state 48) ---")
ca_tx = result[(result["origin_state"] == 6) & (result["dest_state"] == 48)]
print(
    ca_tx[["county_fips", "county_name", "avg_agi_pp_thousands",
           "origin_effective_rate", "rate_diff", "tax_savings_thousands"]]
    .sort_values("tax_savings_thousands", ascending=False)
    .head(5)
    .to_string(index=False)
)

print("\n--- TX counties: savings moving to CA (state 6) — expect negative ---")
tx_ca = result[(result["origin_state"] == 48) & (result["dest_state"] == 6)]
print(
    tx_ca[["county_fips", "county_name", "avg_agi_pp_thousands",
           "origin_effective_rate", "rate_diff", "tax_savings_thousands"]]
    .head(5)
    .to_string(index=False)
)

# ---- Save ----
result.to_csv(OUT_PATH, index=False)
print(f"\nSaved to: {OUT_PATH}")
